File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
'''
1. 아이디어
    - 주어진 정수들을 이용하여 target을 만드는 경우의 수 구하기(조합 탐색: DFS)
    1) 그래프를 그려, 좌로 이동 시에는 음수를 우로 이동 시에는 양수를 표현한다
    2) 모든 그래프의 리프까지 도달한 후, 각 값들이 target과 같다면 cnt += 1
    - graph를 구성해야 한다
'''

def solution(numbers, target):
    
    def dfs(index, current_sum):
        
        if index == len(numbers): # 종료 조건 필요(재귀함수)
            return 1 if current_sum == target else 0
        
        return dfs(index + 1, current_sum + numbers[index]) + \
               dfs(index + 1, current_sum - numbers[index])
    
    return dfs(0, 0)

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
/*
Q. n개의 자연수들을 이용해서 타겟 넘버를 만드세요. 만들 수 있는 모든 경우의 수를 반환하기
A. 모든 경우의 수를 구하는 알고리즘은 보통 DFS로 가지를 뻗어나가는 형태의 그래프 탐색을 이용.
*/

import java.util.*;

class Solution {
    
    static int answer = 0;
    
    public int solution(int[] numbers, int target) {
         
        dfs(numbers, target, 0, numbers[0]); // 양수로 시작
        dfs(numbers, target, 0, -numbers[0]); // 음수로 시작
        
        return answer;
        
    }
    
    private void dfs(int[] numbers, int target, int idx, int total){
        // 종료 조건
        if(idx+1 == numbers.length){
            if(total == target) answer += 1;
            return;
        }
        
        // 재귀
        dfs(numbers, target, idx+1, total + numbers[idx+1]);
        dfs(numbers, target, idx+1, total - numbers[idx+1]);
        
    }
    
}